src/ucp/rndv/proto_rndv.c

```diff
@@ -278,115 +303,134 @@ ucp_proto_rndv_ctrl_init_priv(const ucp_proto_rndv_ctrl_init_params_t *params,
+    perf_elems[num_elems++]     = remote_perf;
+    if (params->unpack_perf != NULL) {
+        perf_elems[num_elems++] = params->unpack_perf;
+    }
 
-void
-ucp_proto_rndv_set_variant_config(const ucp_proto_init_params_t *init_params,
-                                  const ucp_proto_init_elem_t *proto,
-                                  const ucp_proto_select_param_t *select_param,
-                                  void *priv, ucp_proto_config_t *proto_config)
-{
-    proto_config->proto          = ucp_protocols[proto->proto_id];
-    proto_config->priv           = priv;
-    proto_config->ep_cfg_index   = init_params->ep_cfg_index;
-    proto_config->rkey_cfg_index = init_params->rkey_cfg_index;
-    proto_config->select_param   = *select_param;
+    status = ucp_proto_perf_aggregate(ucs_string_buffer_cstr(&perf_name_buf),
+                                      perf_elems, num_elems, &perf);
+    if (status != UCS_OK) {
+        goto out_destroy_remote_perf;
+    } else if (ucp_proto_perf_is_empty(perf)) {
+        /* Empty perf means that this ctrl message cannot be used with that
+         * remote proto (e.g. rndv/rtr/mtype + rndv/send/ppln) */
+        goto out_destroy_perf;
+    }
+
+    ucp_proto_rndv_set_variant_config(&params->super.super, remote_proto,
+                                      remote_select_param, remote_priv,
+                                      &rpriv->remote_proto_config);
+
+    /* Set priority and threshold for this variant */
+    cfg_thresh   = params->super.cfg_thresh;
+    cfg_priority = params->super.cfg_priority;
+    if ((remote_proto->cfg_thresh != UCS_MEMUNITS_INF) &&
+        (remote_proto->cfg_thresh != UCS_MEMUNITS_AUTO)) {
+        /* Consider remote priority and threshold only if RNDV_SCHEME or
+            * BCOPY/ZCOPY thresh are set to force these settings */
+        cfg_priority = remote_proto->cfg_priority;
+        cfg_thresh   = (cfg_thresh == UCS_MEMUNITS_AUTO) ?
+                        remote_proto->cfg_thresh :
+                        ucs_max(cfg_thresh, remote_proto->cfg_thresh);
+    }
+
+    /* Remote variants priorities are used to respect RNDV_SCHEME setting
+     * so they should contain value greater than RTS `cfg_thresh`.
+     * Equality is allowed for RTR remote variants.
+     */
+    ucs_assertv(params->super.cfg_priority <= remote_proto->cfg_priority,
+                "remote_proto=%s params->super.cfg_priority=%u "
+                "remote_proto->cfg_priority=%u", variant_name,
+                params->super.cfg_priority, remote_proto->cfg_priority);
+
+    status = ucp_proto_select_add_proto(&params->super.super, cfg_thresh,
+                                        cfg_priority, perf, rpriv, priv_size);
+    if (status == UCS_OK) {
+        /* `perf` ownership was moved to `init_params->ctx`, so it shouldn't be
+         * destroyed now */
+        goto out_destroy_remote_perf;
+    }
+
+out_destroy_perf:
+    ucp_proto_perf_destroy(perf);
```

Comment:
this looks a bit weird (not following the convention)

Reply:
The logic here is that we shouldn't destroy `perf` structure in case of success because ownership of that function was moved to  `init_params->ctx`. What is unclear here?